src/riak_core_info_service.erl

```diff
@@ -0,0 +1,104 @@
+%% -------------------------------------------------------------------
+%%
+%% Copyright (c) 2017 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%%% @doc
+%%%      The `riak_core_info_service' is a way for dependencies of
+%%%      `riak_core' to be registered to receive messages from
+%%%      `riak_core' without a cyclic dependency graph.
+%%%
+%%%      == Callbacks ==
+%%%      The dependency needs to know to which pid it should send
+%%%      requests for `riak_core' data. The pid will be sent to the
+%%%      `Registration' callback.
+%%%
+%%%      The `Provider' callback is a function inside `riak_core' that returns
+%%%      information that the dependency needs.
+%%%
+%%%      The `Handler' callback is a function in the dependency that
+%%%      expects the reply from the `Provider'.
+%%%
+%%%      === Handler Parameters ===
+%%%      The arguments to `Handler' will be, in order:
+%%%      <ol><li>Each item, if any, provided in the list in the 3rd element of the registration tuple</li>
+%%%      <li>The result from `Provider' wrapped in a 3-tuple:
+%%%        <ol><li>The result</li>
+%%%            <li>The list of parameters passed to `Provider'</li>
+%%%            <li>The opaque `HandlerContext' (see {@section Request Message})</li>
+%%%        </ol></li></ol>
+%%%
+%%%
```

Comment:
Reworking this per discussions with Ted.

Notable change to behavior: `undefined` will no longer be an acceptable callback.